### PR TITLE
net/dns/resolver: add subdomain resolver support in MagicDNS

### LINKED ISSUE #15037: IPv6 DNS on network without IPv6 WAN leads to cyclic dns failures on Android
### What is the issue?

Setup at home uses OpenWRT with its default settings. This sets up a IPv6 DHCPv6-service + IPv6 RA-service.

This lead to a DNS configuration on my Android device as:
```
1.1.1.1
1.0.0.1
fd7b:(...)::1
fd62:(...)::1
```

Running without Tailscale MagicDNS/VPN works correctly and has no DNS failures in this setup.

But once Tailscale MagicDNS and the VPN is turned on, resolving a DNS query will fail in _all_ apps between 1 to 2 times before it succeeds. When using Firefox, this is just refreshing the page twice will let it function again. This consistent nature of needing only to refresh once or twice leads me to belive there is something at fault in how DNS servers get selected by Tailscale.

Disabling DHCPv6+RA and trying again leads to only `1.1.1.1` and `1.0.0.1` DNS servers being available inside the Android settings, and DNS resolution start functioning without failures through Tailscale.

Side note is that on this network, Tailscale will frequently complain in app for being "Out of Sync" as it cannot connect to the coordination server when under this bad configuration. Another is OpenWRT was manually configured with a 6to4 bridge, but was not functioning.

### Steps to reproduce

1. Set up a network with DHCPv6-service + IPv6 RA-service (default) and a non functional WAN6 interface in OpenWRT.
2. Have Tailscale set up with MagicDNS and turn on Tailscale.
3. Browse the web using a webbrowser like Firefox, ensure to try many domain names that are not cached.
4. See cyclic dns errors of sometimes working, refresh until it works again, then errors after that functions, etc.

### Are there any recent changes that introduced the issue?

Only started using Tailscale Android as of January 2025, so this has been here for all my period of using Tailscale Android.

### OS

Android

### OS version

Android 15

### Tailscale version

1.80.0

### Other software

OpenWRT 23.05.5

### Bug report

BUG-041741c1b786651bdb68617a27c6eabcde7d53fb3bf8d8a4927bd7da55dd4af4-20250215110038Z-5a0667cbeed307df

### PR DESCRIPTION
**Pull request by rscastro**
This PR adds processing of subdomains that if there is an entry for a wildcard entry in hosts for a domain, MagicDNS will resolve with the host IP.

Fixes #15037

### CODE REVIEW

File: tailscaled
Comment (raggi) [inline tailscaled:1]: remove this binary! 🤣

File: net/dns/resolver/tsdns.go [hunk 1]
```diff
@@ -196,6 +196,7 @@ type Resolver struct {
 	mu           sync.Mutex
 	localDomains []dnsname.FQDN
 	hostToIP     map[dnsname.FQDN][]netip.Addr
+	wcHostList   []dnsname.FQDN
```
Comment (raggi) [inline net/dns/resolver/tsdns.go:199]: `wc` is a little foreign to me here, I'd find something like `wildcards` or `wildDomains` faster to interpret

File: net/dns/resolver/tsdns.go [hunk 2]
```diff
@@ -596,10 +611,29 @@ func (r *Resolver) resolveLocal(domain dnsname.FQDN, typ dns.Type) (netip.Addr,
 
 	r.mu.Lock()
 	hosts := r.hostToIP
+	wcHosts := r.wcHostList
```
Comment (raggi) [inline net/dns/resolver/tsdns.go:614]: similarly here I'd probably just call this `wildcards`

File: net/dns/resolver/tsdns.go [hunk 3]
```diff
@@ -596,10 +611,29 @@ func (r *Resolver) resolveLocal(domain dnsname.FQDN, typ dns.Type) (netip.Addr,
...
 	r.mu.Unlock()
 
-	addrs, found := hosts[domain]
+	var found = false
+	var addrs []netip.Addr
+
+	addrs, found = hosts[domain]
+	if !found {
+		// Find matching subdomain entries
+		for _, key := range wcHosts {
```
Comment (raggi) [inline net/dns/resolver/tsdns.go:624]: key is `wildcard`

File: net/dns/resolver/tsdns.go [hunk 4]
```diff
@@ -596,10 +611,29 @@ func (r *Resolver) resolveLocal(domain dnsname.FQDN, typ dns.Type) (netip.Addr,
...
+			// Entries in wcHosts already have a trailing '.', and Contains adds another
```
Comment (raggi) [inline net/dns/resolver/tsdns.go:625]: you can skip `ToFQDN` as `key` is already an fqdn

**Review by raggi — COMMENTED** [review]
Looking good overall, definitely remove the binary, and most of the rest of the comments are naming related.

File: net/dns/resolver/tsdns.go (hunk 4 above)
Comment (rscastro) [reply net/dns/resolver/tsdns.go:625]: I have to convert it to a string to remove the '.' then back to an FQDN

Comment (raggi) [reply net/dns/resolver/tsdns.go:626]: Are you sure, doesn't ToFQDN also perform this removal?

**Comment by bradfitz** [comment]
`Fixes #15037` doesn't reference an issue in this repo.

Do you need some new docs in tailcfg.go to document these new semantics on wildcards under DNSConfig? Does it need a capver bump to signal to control that the client supports wildcards? (I don't know)

File: net/dns/resolver/tsdns.go (hunk 4 above)
Comment (rscastro) [reply net/dns/resolver/tsdns.go:626]: Sorry I think I'm not understanding the change here. The original problem was running `Contains` off of `wilcard` since it adds a trailing '.'. So I i did remove the `[1:]` since ToFQDN does remove the '.'

**Comment by rscastro** [comment]
@bradfitz are there any other spots where you think I'm missing or need changes in documentation?

File: net/dns/resolver/tsdns.go (hunk 4 above)
Comment (raggi) [reply net/dns/resolver/tsdns.go:627]: you should be able to ensure that wildcards contains FQDN values pre-sanitized in whatever way they need to be consumed here, so that no parsing or reconstruction needs to happen when answering a query.

I'm a little confused by the discussion of trailing periods and the slice prefix trim, particularly because the type has interest in both - I'm not sure which you're actually discussing.

Either way, do the necessary construction when wildcards is populated so here all you need to do during a query is check for a match.

File: net/dns/resolver/tsdns.go [hunk 5]
```diff
@@ -246,6 +248,21 @@ func (r *Resolver) SetConfig(cfg Config) error {
 	r.localDomains = cfg.LocalDomains
 	r.hostToIP = cfg.Hosts
 	r.ipToHost = reverse
+
+	// Extract wildcard hosts
+	keys := make([]dnsname.FQDN, 0)
+	for k := range r.hostToIP {
+		if len(k) > 0 && k[0] == '.' {
```
Comment (bradfitz) [inline net/dns/resolver/tsdns.go:255]: `if strings.HasPrefix(k, ".") {` ?

File: net/dns/resolver/tsdns.go [hunk 6]
```diff
@@ -596,10 +613,29 @@ func (r *Resolver) resolveLocal(domain dnsname.FQDN, typ dns.Type) (netip.Addr,
...
 	hosts := r.hostToIP
+	wildcards := r.wildcards
 	localDomains := r.localDomains
 	r.mu.Unlock()
 
-	addrs, found := hosts[domain]
+	var found = false
+	var addrs []netip.Addr
+
+	addrs, found = hosts[domain]
```
Comment (bradfitz) [inline net/dns/resolver/tsdns.go:623]: need the var lines? Can you just do `addrs, found := hosts[domain]` ?

File: net/dns/resolver/tsdns.go [hunk 7]
```diff
@@ -596,10 +613,29 @@ func (r *Resolver) resolveLocal(domain dnsname.FQDN, typ dns.Type) (netip.Addr,
...
+	if !found {
+		// Find matching subdomain entries
+		for _, wildcard := range wildcards {
+			// Entries in wcHosts already have a trailing '.', and Contains adds another
+			fqdn, err := dnsname.ToFQDN(string(wildcard))
```
Comment (bradfitz) [inline net/dns/resolver/tsdns.go:628]: we shouldn't need to be doing any allocation at serving time here.

let's get it into a form in memory that's efficient to serve from. The O(n) loop over wildcards is already slightly shady (because I don't know how many wildcards users will end up having... 0-few usually, but what's the upper bound? hundreds? thousand or more?)

File: net/dns/resolver/tsdns.go [hunk 8]
```diff
@@ -246,6 +249,15 @@ func (r *Resolver) SetConfig(cfg Config) error {
 	r.localDomains = cfg.LocalDomains
 	r.hostToIP = cfg.Hosts
 	r.ipToHost = reverse
+
+	// Extract wildcard hosts
+	keys := make(map[dnsname.FQDN][]netip.Addr)
```
Comment (bradfitz) [inline net/dns/resolver/tsdns.go:254]: In the common case we won't need this map so I'd just declared it here with its nil zero value and not use make.  You can use mak.Set in the loop to lazily set it.

File: net/dns/resolver/tsdns.go [hunk 9]
```diff
@@ -596,11 +608,23 @@ func (r *Resolver) resolveLocal(domain dnsname.FQDN, typ dns.Type) (netip.Addr,
...
 	hosts := r.hostToIP
+	wildcards := r.wildcards
 	localDomains := r.localDomains
 	r.mu.Unlock()
 
-	addrs, found := hosts[domain]
-	if !found {
+	addrs, ok := hosts[domain]
+	if !ok {
+		labels := strings.Split(string(domain), ".")
```
Comment (bradfitz) [inline net/dns/resolver/tsdns.go:617]: strings.Split allocates. As does strings.Join.

We should be able to do this by looking for dots and slicing the input string?

Add an allocation test too with testing.AllocsPerRun?

File: net/dns/resolver/tsdns.go [hunk 10]
```diff
@@ -596,11 +609,24 @@ func (r *Resolver) resolveLocal(domain dnsname.FQDN, typ dns.Type) (netip.Addr,
...
+	if !ok {
+		// Look for segment in map when '.' is found
+		for i := 0; i < len(domain); i++ {
+			if domain[i] == '.' {
+				segment := domain[i:]
+				if addrs, ok = wildcards[segment]; ok {
+					break
+				}
+			}
+		}
```
Comment (maisem) [inline net/dns/resolver/tsdns.go:626]: you can do the same w/o using an extra map

```
		d := domain.WithTrailingDot()
		for ix := strings.IndexRune(d, '.'); ix >= 0; ix = strings.IndexRune(d, '.') {
			h := dnsname.FQDN(d[ix:]) // include the dot
			d = d[ix+1:]
			if addrs, ok = hosts[h]; ok {
				break
			}
		}
```

**Review by maisem — COMMENTED** [review]
I think you would also need to change [dnsConfigFromNetmap](https://github.com/tailscale/tailscale/blob/a7e4cebb90e3c16dc8dc9e511daf3900c0135f79/ipn/ipnlocal/local.go#L3383-L3386). Currently it [strips](https://github.com/tailscale/tailscale/blob/a7e4cebb90e3c16dc8dc9e511daf3900c0135f79/util/dnsname/dnsname.go#L28-L30) the leading dot

File: util/dnsname/dnsname.go [hunk 11]
```diff
@@ -20,14 +20,12 @@ const (
 // A FQDN is a fully-qualified DNS name or name suffix.
 type FQDN string
 
-func ToFQDN(s string) (FQDN, error) {
+// Retains leading '.' on FQDN to populate Hosts with wildcard entries from ExtraRecords
```
Comment (bradfitz) [inline util/dnsname/dnsname.go:23]: See https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences for Go doc style. It should be a complete sentence starting with the name of the func and ending in a period.

But this is a util package that isn't about tailcfg or even Tailscale so it can't mention things like ExtraRecords which is a very Tailscale wire protocol concept.

File: net/dns/resolver/tsdns.go [hunk 12]
```diff
@@ -71,7 +72,9 @@ type Config struct {
 	// Queries only match the most specific suffix.
 	// To register a "default route", add an entry for ".".
 	Routes map[dnsname.FQDN][]*dnstype.Resolver
-	// LocalHosts is a map of FQDNs to corresponding IPs.
+	// Hosts maps either domain names or wildcards to IP(s).
+	// If the map key begins with a dot (e.g. ".foo.com.") then it's considered a wildcard,
+	// matching *.foo.com.
```
Comment (raggi) [inline net/dns/resolver/tsdns.go:77]: would be nice to explicitly say whether or not foo.com is matched by .foo.com here.

File: util/dnsname/dnsname.go [hunk 13]
```diff
@@ -20,14 +20,12 @@ const (
 // A FQDN is a fully-qualified DNS name or name suffix.
 type FQDN string
 
-func ToFQDN(s string) (FQDN, error) {
+// ToFQDNSuffix converts a string to a FQDN that retains any leading '.' in the case of wildcards.
+func ToFQDNSuffix(s string) (FQDN, error) {
```
Comment (raggi) [inline util/dnsname/dnsname.go:24]: Should this explicitly add a leading `.` if there isn't one? If the user intent is "I want a suffix", then we probably should. We could also make the doc string more exact in this regard - do you always get a suffix matching FQDN, or sometimes you might get just a fully qualified domain name?

File: util/dnsname/dnsname.go [hunk 14]
```diff
@@ -65,6 +60,17 @@ func ToFQDN(s string) (FQDN, error) {
 	return FQDN(raw), nil
 }
 
+func ToFQDN(s string) (FQDN, error) {
```
Comment (raggi) [inline util/dnsname/dnsname.go:63]: Add a documentation string here that expresses that this explicitly strips a leading '.' and thereby prevents the creation of suffix matches.

File: util/dnsname/dnsname.go [hunk 15]
```diff
@@ -41,18 +39,15 @@ func ToFQDN(s string) (FQDN, error) {
...
+		// Ignore leading '.' from wildcards for label processing
+		if i == 0 && s[i] == '.' {
+			st = i + 1
+			continue
+		}
 		if s[i] != '.' {
 			continue
 		}
 		label := s[st:i]
-		// You might be tempted to do further validation of the
```
Comment (raggi) [inline util/dnsname/dnsname.go:48]: I'm not sure if this was intentional, we should keep some amount of it at least, it's reasonable advice.

File: util/dnsname/dnsname.go [hunk 16]
```diff
@@ -65,6 +60,17 @@ func ToFQDN(s string) (FQDN, error) {
...
 
+func ToFQDN(s string) (FQDN, error) {
+	if len(s) == 0 || s == "." {
+		return FQDN("."), nil
+	}
+
+	if s[0] == '.' {
+		s = s[1:]
+	}
+	return ToFQDNSuffix(s)
```
Comment (raggi) [inline util/dnsname/dnsname.go:71]: I see this is taking on the role of a general constructor. I think the API would be more clear if we had:

NewFQDN  // creates an FQDN without validating leading '.' presence, it accepts with or without. It adds a trailing '.' if one is not present.
ToFQDN   // ensures that there is no leading '.', removing it if found, otherwise delegates to NewFQDN.
ToFQDNSuffix // ensures that there is a leading '.', adding it if not found, otherwise delegates to NewFQDN.

The only problem currently is that ToFQDN has opinionated behavior (stripping '.'), but ToFQDNSuffix has unopinionated behavior (ignoring leading '.' presence), but if you were to just make assumptions based on the names, it would seem like the more specific name `ToFQDNSuffix` would have a more specific behavior.

**Review by raggi — APPROVED** [review]
I have some suggestions, but the approach and so on look good to me.

**Review by raggi — APPROVED** [review]

File: net/dns/resolver/tsdns.go [hunk 17]
```diff
@@ -71,7 +72,10 @@ type Config struct {
 	// Queries only match the most specific suffix.
 	// To register a "default route", add an entry for ".".
 	Routes map[dnsname.FQDN][]*dnstype.Resolver
-	// LocalHosts is a map of FQDNs to corresponding IPs.
+	// Hosts maps either domain names or wildcards to IP(s).
+	// If the map key begins with a dot (e.g. ".foo.com.") then it's considered a wildcard,
+	// matching *.foo.com. "foo.com" would not get matched with ".foo.com" since the label "foo"
```
Comment (bradfitz) [inline net/dns/resolver/tsdns.go:77]: this second sentence is still about wild cards but we never gave an example of the normal form of this map key (non-wildcard). Include a little example first of "record.foo.com." (with the trailing dot) so it doesn't look like "foo.com" (which doesn't have a trailing dot in that example) is meant as a map key?

File: util/dnsname/dnsname.go [hunk 18]
```diff
@@ -20,14 +20,12 @@ const (
 // A FQDN is a fully-qualified DNS name or name suffix.
 type FQDN string
 
-func ToFQDN(s string) (FQDN, error) {
+// NewFQDN converts a string to a FQDN that retains any leading '.' in the case of wildcards.
```
Comment (bradfitz) [inline util/dnsname/dnsname.go:23]: This package should have no concept of wildcards or our internal/wire representation of them.

File: util/dnsname/dnsname.go [hunk 19]
```diff
@@ -65,6 +68,30 @@ func ToFQDN(s string) (FQDN, error) {
 	return FQDN(raw), nil
 }
 
+// ToFQDN strips a leading '.' on a string before converting to a FQDN.
+func ToFQDN(s string) (FQDN, error) {
+	if len(s) == 0 || s == "." {
+		return FQDN("."), nil
```
Comment (bradfitz) [inline util/dnsname/dnsname.go:74]: The real problem is the FQDN type has no docs. But then these docs read contradictorily: it says it strips a leading dot but then these two paths return it with a leading dot. But if we do what an "FQDN" was (the undocumented type) then perhaps it'd make sense.

File: net/dns/resolver/tsdns_test.go [hunk 20]
```diff
@@ -373,6 +384,14 @@ func TestResolveLocal(t *testing.T) {
 			if ip != tt.ip {
 				t.Errorf("ip = %v; want %v", ip, tt.ip)
 			}
+
+			allocs := testing.AllocsPerRun(1000, func() {
+				r.resolveLocal(tt.qname, tt.qtype)
+			})
+			// 4via6 paths have 1 alloc, the rest should be 0
+			if allocs > 1 {
```
Comment (bradfitz) [inline net/dns/resolver/tsdns_test.go:392]: add an alloc test without 4via6 (with just wildcard matching) to see that it's zero
